### PR TITLE
Add offline payment flow

### PR DESCRIPTION
**Pull request by Raibo**
https://github.com/happy-travel/agent-app-project/issues/111

### CODE REVIEW

File: Api/Controllers/AgentControllers/BookingsController.cs [hunk 1]
```diff
@@ -91,6 +91,27 @@ public async Task<IActionResult> Book([FromBody] AccommodationBookingRequest req
         }
 
 
+        /// <summary>
+        ///     Creates booking in one step. Must be payed offline before deadline.
+        /// </summary>
+        /// <param name="request"></param>
+        /// <returns></returns>
+        [HttpPost("book-by-offline")]
```
Comment (DFSko) [inline Api/Controllers/AgentControllers/BookingsController.cs:99]: book by offline? maybe like book by bank transfer?

Comment (mingazhev) [reply Api/Controllers/AgentControllers/BookingsController.cs:99]: Needs discussion, I've wrote my suggestions in discord

File: Api/Services/Accommodations/Bookings/BookingExecution/Flows/IOfflinePaymentBookingFlow.cs [hunk 2]
```diff
@@ -0,0 +1,13 @@
+using System.Threading.Tasks;
+using CSharpFunctionalExtensions;
+using HappyTravel.Edo.Api.Models.Agents;
+using HappyTravel.Edo.Api.Models.Bookings;
+
+namespace HappyTravel.Edo.Api.Services.Accommodations.Bookings.BookingExecution.Flows
+{
+    public interface IOfflinePaymentBookingFlow
+    {
+        Task<Result<AccommodationBookingInfo>> BookByOffline(AccommodationBookingRequest bookingRequest,
```
Comment (mingazhev) [inline Api/Services/Accommodations/Bookings/BookingExecution/Flows/IOfflinePaymentBookingFlow.cs:10]: Just Book will be enough, flow name already contains it's name

File: Api/Services/Accommodations/Bookings/BookingExecution/Flows/OfflinePaymentBookingFlow.cs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+            _requestExecutor = requestExecutor;
+        }
+        
+
+        public Task<Result<AccommodationBookingInfo>> BookByOffline(AccommodationBookingRequest bookingRequest,
+            AgentContext agentContext, string languageCode, string clientIp)
+        {
+            return GetCachedAvailability(bookingRequest)
+                .Ensure(IsPaymentMethodAllowed, "Payment method is not allowed")
+                .Ensure(IsNotDeadlinePassed, "Deadline already passed, can not book")
```
Comment (mingazhev) [inline Api/Services/Accommodations/Bookings/BookingExecution/Flows/OfflinePaymentBookingFlow.cs:40]: IsDeadlineNotPassed will sound better wouldn't it?

File: Api/Services/Accommodations/Bookings/BookingExecution/Flows/OfflinePaymentBookingFlow.cs [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+                .Check(GenerateInvoice)
+                .Bind(SendSupplierRequest)
+                .Bind(GetAccommodationBookingInfo);
+
+
+            bool IsNotDeadlinePassed(BookingAvailabilityInfo bookingAvailability)
+            {
+                var deadlineDate = bookingAvailability.RoomContractSet.Deadline.Date;
+                var dueDate = deadlineDate == null || deadlineDate == DateTime.MinValue ? bookingAvailability.CheckInDate : deadlineDate.Value;
+                return dueDate > _dateTimeProvider.UtcToday();
```
Comment (mingazhev) [inline Api/Services/Accommodations/Bookings/BookingExecution/Flows/OfflinePaymentBookingFlow.cs:51]: I think we should add at least a day there, before this task will be finished
https://happytravel.atlassian.net/browse/NIJO-1305

File: Api/Services/Accommodations/Bookings/BookingExecution/Flows/OfflinePaymentBookingFlow.cs (hunk 3 above)
Comment (Raibo) [reply Api/Services/Accommodations/Bookings/BookingExecution/Flows/OfflinePaymentBookingFlow.cs:40]: IsNotDeadlinePassed lets you know immediately that there is "Not", while IsDeadlineNotPassed could have more chances to be misunderstood.

File: Api/Services/Accommodations/Bookings/BookingExecution/Flows/OfflinePaymentBookingFlow.cs (hunk 4 above)
Comment (Raibo) [reply Api/Services/Accommodations/Bookings/BookingExecution/Flows/OfflinePaymentBookingFlow.cs:51]: I thought that in availability deadline is already adjusted, -1 to -3 days depending on contract kind.
Is it?

File: Api/Controllers/AgentControllers/BookingsController.cs (hunk 1 above)
Comment (Raibo) [reply Api/Controllers/AgentControllers/BookingsController.cs]: changed to `book-for-offline`

File: Api/Services/Accommodations/Bookings/BookingExecution/Flows/OfflinePaymentBookingFlow.cs (hunk 4 above)
Comment (Raibo) [reply Api/Services/Accommodations/Bookings/BookingExecution/Flows/OfflinePaymentBookingFlow.cs]: After discord discussion, added 3 day shift.
When more information appears, we'll do something with this.

File: Api/Services/Accommodations/Bookings/BookingExecution/Flows/OfflinePaymentBookingFlow.cs (hunk 3 above)
Comment (kirillta) [reply Api/Services/Accommodations/Bookings/BookingExecution/Flows/OfflinePaymentBookingFlow.cs:39]: I prefer more natural English in names

**Review by mingazhev — APPROVED** [review]
Waiting for @kirillta approve for deadline buffer

**Review by kirillta — APPROVED** [review]

**Pull request merged**